Upgrade Guides 0.19, 1.0 and Pricing Guide (#701)
liveblocks/liveblocks: docs/pages/guides/upgrading/1.0.mdx

```bash
npm update @liveblocks/client@1.0 @liveblocks/node@1.0
```

The major release of 1.0 marks the maturity of Liveblocks. We added additional

> features to Webhooks, enhancements to the `Room` class, and more. The primary

Reviewer: ```suggestion
features to [Webhooks](/docs/guides/webhooks), enhancements to the `Room` class, and more. The primary
```

### Adds throttle option to createClient [#throttle]

You can now specify the `throttle` option in
[`createClient`](https://liveblocks.io/docs/api-reference/liveblocks-client#createClientThrottle)
that can go as low as 16ms. If not set, the default value is 100ms.

```ts highlight="5"
import { createClient } from "@liveblocks/client";

const client = createClient({
  /* ... other options ... */
  throttle: 16,
});
```

### Improvements to the Room class [#room]

We added new methods to the `Room` class that you can use to obtain the storage
status of a room, subscribe to storage status changes, or reconnect to a room
whenever necessary. You can also use the
[Liveblocks DevTools extension](/devtools) to visualize real-time changes to
storage while implementing these changes.

#### Room.getStorageStatus [#getStorageStatus]

The different storage statuses are:

- `not-loaded`: Initial state when entering the room.
- `loading`: Once the storage has been requested via room.getStorage().
- `synchronizing`: When some local updates have not been acknowledged by
  Liveblocks servers.
- `synchronized`: Storage is in sync with Liveblocks servers.

#### Room.subscribe [#subscribe]

> Use `Room.`subscribe` to subscribe to storage status changes. This method

Reviewer: ```suggestion
Use `Room.subscribe` to subscribe to storage status changes. This method
```

returns an unsubscribe function.

```ts
room.subscribe("storage-status", (status) => {
  // Implement your logic here
  switch (status) {
    case "not-loaded":
      break;
    case "loading":
      break;
    case "synchronizing":
      break;
    case "synchronized":
      break;
    default:
      break;
  }
});
```

#### Room.reconnect [#reconnect]

You can use `room.reconnect()` to close the room connection and try to
[reconnect](/docs/api-reference/liveblocks-client#Room.reconnect).

## Changes in @liveblocks/node [#node]

### Update the authorize option [#authorize]

> We have updated the `authorize` option to make `userId` mandatory. This change

Reviewer: ```suggestion
We have updated the `authorize` method to make `userId` mandatory. This change
```

ties into our new [pricing model](https://liveblocks.io/pricing), which is based

> on Monthly Active Users instead of connections. We use `userId` to track MAU

Reviewer: ```suggestion
on Monthly Active Users (MAU) instead of connections. We use `userId` to track MAU
```

associated with a Liveblocks account.

> If you do not use the `authorize` option, a cookie will be set on the client to

Reviewer: ```suggestion
If you do not use the `authorize` method, a cookie will be set on the client to
```

track the user. This cookie will be used to track MAU associated with a
Liveblocks account and will expire after 30 days.

> ### Webhooks Enhancements [#webhooks]

Reviewer: ```suggestion
### Webhooks enhancements [#webhooks]
```

> We have added two new events to our Webhook functionality: `RoomCreatedEvent`

Reviewer: ```suggestion
We have added two new events to our Webhooks functionality: `RoomCreatedEvent`
```

and `RoomDeletedEvent`. We have also added a `WebhookHandler` class to make it

> easier to verify event requests from Liveblock's Webhook functionality. It also

Reviewer: ```suggestion
easier to verify event requests from Liveblock's Webhooks functionality. It also
```

> provides fully typed `WebhookEvents.`

Reviewer: ```suggestion
provides fully typed `WebhookEvents`.
```

```bash
npm update @liveblocks/client@1.0 @liveblocks/react@1.0 @liveblocks/node@1.0
```

### Step 2: Update your authorize option [#step2]

```ts highlight="20"
import { authorize } from "@liveblocks/node";

// Replace this key with your secret key provided at
// https://liveblocks.io/dashboard/projects/{projectId}/apikeys
const secret = "sk_prod_xxxxxxxxxxxxxxxxxxxxxxxx";

export default async function auth(req, res) {
  /**
   * Implement your own security here.
   *
   * It's your responsibility to ensure that the caller of this endpoint
   * is a valid user by validating the cookies or authentication headers
   * and that it has access to the requested room.
   */
  const room = req.body.room;
  const response = await authorize({
    room,
    secret,
    // Corresponds to the UserMeta[id] type defined in liveblocks.config.ts
    userId: "123", // Required
    groupIds: ["456"], // Optional
    userInfo: {
      // Optional, corresponds to the UserMeta[info] type defined in liveblocks.config.ts
      name: "Ada Lovelace",
      color: "red",
    },
  });
  return res.status(response.status).end(response.body);
}
```

That’s it!

> If you have issues with these new patterns and need help, please
> [let us know](/contact)! We’re here to help!

Reviewer: ```suggestion
If you have issues with these new patterns and need help, please let us know [by email](mailto:support@liveblocks.io), or by joining our [Discord community](/discord)! We’re here to help!
```

The `contact` page is too generic, and we're missing an occasion to make people join our Discord server

---
meta:
  title: "Upgrading to 1.0"
  parentTitle: "Guides"
  description: "Guide to upgrade to Liveblocks version 1.0"
---

```bash
npm update @liveblocks/client@1.0 @liveblocks/node@1.0
```

The major release of 1.0 marks the maturity of Liveblocks. We added additional
features to [Webhooks](/docs/guides/webhooks), enhancements to the `Room` class,
and more. The primary change you should note is the transition to making the
`userId` mandatory in the `authorize` option. This change is in line with our
[new pricing model](https://liveblocks.io/pricing).

Let’s take a look at the changes below!

## Changes in @liveblocks/client [#client]

### Adds throttle option to createClient [#throttle]

> You can now specify the `throttle` option in

Reviewer 2: If I'm not mistaken, it was already possible to specify the throttle value in the createClient option. The difference is that it's now possible to go as low as 16ms.

Reviewer: True!

Reviewer 3: @ofoucherot is right. The change here is that you can go as low at 16ms, meaning you can theoretically get 60 FPS updates. Great for live cursors!

@CTNicholas had some excellent tweets about this.

Can we please change the heading and description to reflect that?

Reviewer 4: Here's the [tweet](https://twitter.com/ctnicholasdev/status/1622656493769359360), previously you could go as low as 80

Author: Thanks @CTNicholas I added the tweet to the guide,  as well as link to example.

- `synchronizing`: When some local updates have not been acknowledged by
  Liveblocks servers.
- `synchronized`: Storage is in sync with Liveblocks servers.

#### Room.subscribe [#subscribe]

Use `Room.subscribe` to subscribe to storage status changes. This method returns
an unsubscribe function.

```ts
room.subscribe("storage-status", (status) => {
  // Implement your logic here
  switch (status) {
    case "not-loaded":
      break;
    case "loading":
      break;
    case "synchronizing":
      break;
    case "synchronized":
      break;
    default:
      break;
  }
});
```

#### Room.reconnect [#reconnect]

You can use `room.reconnect()` to close the room connection and try to
[reconnect](/docs/api-reference/liveblocks-client#Room.reconnect).

## Changes in @liveblocks/node [#node]

### Update the authorize option [#authorize]

We have updated the `authorize` method to make `userId` mandatory. This change
ties into our new [pricing model](https://liveblocks.io/pricing), which is based
on Monthly Active Users (MAU) instead of connections. We use `userId` to track
MAU associated with a Liveblocks account.

> If you do not use the `authorize` function, a cookie will be set on the client

Reviewer 2: I would specify:

> If you do not use the `authorize` function but call the authorize with secret key API endpoint directly, you should pass the `userId` even if it's not mandatory on the API level. (Because it's used for the MAU etc....)

> If you do not use the `authorize` function because you are using the public API key method, a cookie will etc...

`Liveblocks/node` is a SDK with util methods to use our REST API. Currently, there is only a method for the authorize endpoint but we plan on adding all endpoints.

Some customers can still use the secret API key authentication method, but by calling the endpoint directly.

to track the user. This cookie will be used to track MAU associated with a
Liveblocks account and will expire after 30 days.

### Webhooks enhancements [#webhooks]

We have added two new events to our Webhooks functionality: `RoomCreatedEvent`
and `RoomDeletedEvent`. We have also added a `WebhookHandler` class to make it
easier to verify event requests from Liveblock’s Webhooks functionality. It also
provides fully typed `WebhookEvents`.

```ts
import { WebhookHandler } from "@liveblocks/node";

const webhookHandler = new WebhookHandler(process.env.SECRET);

const event = webhookHandler.verifyRequest({
  headers: req.headers,
  rawBody: req.body,
});
```

Verification of events is critical to ensure that the events are coming from
Liveblocks and not from a malicious source. Check out our
[Webhooks guide](https://liveblocks.io/docs/guides/webhooks) for more details.

## Changes for all Liveblocks packages [#all]

> When upgrading to 1.0, the primary breaking change to consider is updating the

Reviewer 2: Why is this part under `Changes for all Liveblocks packages`?
The change is only for `liveblocks/node`.

`authorize` option to include the now-mandatory `userId` field if you use
`@liveblocks/node`. This change reflects our new pricing model based on Monthly
Active Users instead of connections.

### Step 1: Update your Liveblocks packages [#step1]

```bash
npm update @liveblocks/client@1.0 @liveblocks/react@1.0 @liveblocks/node@1.0
```

> ### Step 2: Update your authorize option [#step2]

Reviewer 2: 🎉

---
meta:
  title: "Upgrading to 1.0"
  parentTitle: "Guides"
  description: "Guide to upgrade to Liveblocks version 1.0"
---

```bash

> npm update @liveblocks/client@1.0 @liveblocks/node@1.0

Reviewer 3: Should the upgrade instruction be introduce with a short paragraph?

To update to Liveblocks version 1.0, run the following command using your preferred package manager:

[insert command line here]

Plus, we could do this at the beginning of each section for the each package:

To update to @liveblocks/client to 1.0, run the following command using your preferred package manager:

[insert command line here for @liveblocks/client]

> npm update @liveblocks/client@1.0 @liveblocks/node@1.0

Reviewer 3: Here is what Next does: https://nextjs.org/docs/upgrading

I find that pretty clear to follow: 
- intro paragraph
- command to upgrade
- version changes summary

That format makes it super easy for engineers to see what's changed at glance and what will be impacted / aka how hard the upgrade will be.

The goal is make it as obvious as possible the amount of effort the upgrade will be for people.

> npm update @liveblocks/client@1.0 @liveblocks/node@1.0

Author: Format sounds good to me, happy to update.  I updated the other guides, should be `npm install <@package-name/@version>`

> npm update @liveblocks/client@1.0 @liveblocks/node@1.0

Author: Format sounds good to me, happy to update.  I updated the other guides, should be `npm install <@package-name/@version>`

> npm update @liveblocks/client@1.0 @liveblocks/node@1.0

Author: If we want to pattern match off of the NextJS summary, I could change the "Impovements" section to Summary, and make it have less examples.  Let me know what you think.

```

> The major release of 1.0 marks the maturity of Liveblocks. The primary change

Author: Sounds good.  Will update.  I pulled from CHANGELOG, so let me know if you want me to keep same there.

you should note is the transition to making the `userId` mandatory in the
`authorize` option. This change is in line with our
[new pricing model](https://liveblocks.io/pricing).

Let’s take a look!

> ## Upgrading to Liveblocks 1.0

Reviewer 3: Do we really need this title here? Isn't this the title of the page already?

Author: I thought it would be nice to have in the side nav to separate improvements/ changes from upgrade steps.  I can remove, or rename to `Upgrade Steps` or `Package Specific Upgrade Steps`
